homeassistant/components/usb/__init__.py

```diff
@@ -400,15 +447,27 @@ async def _async_process_ports(self, ports: Sequence[ListPortInfo]) -> None:
                 if dev.device.startswith("/dev/cu.SLAB_USBtoUART")
             }
 
-            usb_devices = [
+            usb_devices = {
                 dev
                 for dev in usb_devices
                 if dev.serial_number not in silabs_serials
                 or (
                     dev.serial_number in silabs_serials
                     and dev.device.startswith("/dev/cu.SLAB_USBtoUART")
                 )
-            ]
+            }
+
+        added_devices = usb_devices - self._last_processed_devices
+        removed_devices = self._last_processed_devices - usb_devices
+        self._last_processed_devices = usb_devices
+
+        _LOGGER.debug(
+            "Added devices: %r, removed devices: %r", added_devices, removed_devices
+        )
+
+        if added_devices or removed_devices:
+            for callback in self._port_event_callbacks.copy():
+                callback(added_devices, removed_devices)
```

Comment:
We might want to suppress exceptions and log them from the callback so that the fail callback doesn't affect the rest of this function as we are running external code here